docs/signing-tool-spec.md

```diff
@@ -0,0 +1,129 @@
+- Support a single certificate for all files in the operation.
+
+**Non-Goals**
+
+- Strong Name signing won't be in v1; guidance is to use an snk not based on a cert. If easy, perhaps can revisit.
+- Support for Microsoft internal builds and signing services.
+- Containers, including Notary v2 support.
+- Extensibility. v1 will support different signing providers.
+- Support for platforms other than Windows x64. Future work will be required to support ARM64 and non-Windows hosts. Support for certain file types may be limited due to platform support.
+- Offline distribution.
+
+#### v1
+
+**Goals**
+
+- Extensibility mechanism to support different code signing providers with a dynamic lookup so the core client remains agnostic of the backend
+- Offline distribution for core plus backend provider
+- Three providers: Certificate Store, Azure Key Vault, Azure Code Signing
+- Support for additional formats: [.HLKX](https://github.com/dotnet/sign/issues/422), [VBA](https://github.com/dotnet/sign/issues/364)
+- Verification of signatures
+
+
+### Inputs
+
+The parameters to the signing client are as follows:
+
+`sign code AzureKeyVault [options] <files>`
+
+Options:
+
+- `-o | --output` - The output file. If omitted, overwrites input. Must be a directory if multiple files are specified.
+- `-b | --base-directory` - Base directory for files to override the working directory.
+- `-f | --force` - Overwrites a signature if it exists.
+- `-m | --max-concurrency` - Maximum concurrency (default is 4)
+- `-fl | --filelist` - Path to file containing paths of files to sign within an archive
+- `-fd | --file-digest` - The digest algorithm to hash the file with.
+- `-tr | --timestamp-rfc3161` - Specifies the RFC 3161 timestamp server's URL.
+- `-d | --description` - Description
+- `-u | --descriptionUrl` - Description Url
+- `-td | --timestamp-digest` - Used with the -tr switch to request a digest algorithm used by the RFC 3161 timestamp server.
+- `-kvu | --azure-key-vault-url` - The URL to an Azure Key Vault.
+- `-kvt | --azure-key-vault-tenant-id` - The Tenant Id to authenticate to the Azure Key Vault..
+- `-kvi | --azure-key-vault-client-id` - The Client ID to authenticate to the Azure Key Vault.
+- `-kvs | --azure-key-vault-client-secret` - The Client Secret to authenticate to the Azure Key Vault.
+- `-kvc | --azure-key-vault-certificate` - The name of the certificate in Azure Key Vault.
+- `-kvm | --azure-key-vault-managed-identity` - Use a Managed Identity to access Azure Key Vault.
+
+Signing an archive type (`.zip`, `.nupkg`, `.vsix`) will open up the archive and sign any supported file types. It is strongly recommended to use the `filter` parameter to explicitly list the files inside the archive that should be signed. Signing is recursive; it will sign contents of any detectected `Zip`, `NuPkg` or `VSIX` files inside the uploaded one. After signing contents of the archive, the archive itself is signed if supported (currently `VSIX`).
```

Comment:
I'm curious how the "archive type" is detected. Magic bytes at the beginning of the file? File extension?

Reply:
It's a list of specific types--you can see them here: https://github.com/dotnet/sign/blob/cli/src/Sign.Core/Containers/ContainerProvider.cs#L36-L60

Reply:
I see .snupkg on there, NuGet.org does not verify .snupkg signatures today and instead depends on the .snupkg PDB matching the .nupkg DLLs. The consumer's debugger does not fetch the .snupkg signature so the benefit is not clear. So I think having this tool sign .snupkg files is a bit strange and is likely unnecessary.

Reply:
It was included for completeness as snupkg is just another NuGet package.

Reply:
Okay, should it be called out in the spec?

Having the tool signing something that no one verifies could be confusing to the user.